ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/HiveRowIsDeletedPropagator.java

```diff
@@ -17,72 +17,139 @@
- * General rule: we expect that the rowIsDeleted column is the last column in the input rowType of the current
+ * {@link ReflectiveVisitor} to propagate row is deleted or inserted columns to all HiveRelNodes' rowType in the plan.
+ * General rule: we expect that these columns are the last columns in the input rowType of the current
  * {@link org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveRelNode}.
+ *
+ * This class is part of incremental rebuild of materialized view plan generation.
+ * <br>
+ * @see AlterMaterializedViewRebuildAnalyzer
+ * @see HiveAggregateInsertDeleteIncrementalRewritingRule
  */
-public class HiveRowIsDeletedPropagator extends HiveRelShuttleImpl {
+public class HiveRowIsDeletedPropagator implements ReflectiveVisitor {
 
   protected final RelBuilder relBuilder;
+  protected final Deque<RelNode> stack = new ArrayDeque<>();
+  private final ReflectUtil.MethodDispatcher<RelNode> dispatcher;
 
   public HiveRowIsDeletedPropagator(RelBuilder relBuilder) {
     this.relBuilder = relBuilder;
+    this.dispatcher = ReflectUtil.createMethodDispatcher(
+        RelNode.class, this, "visit", RelNode.class, Context.class);
   }
 
   public RelNode propagate(RelNode relNode) {
-    return relNode.accept(this);
-  }
-
-  /**
-   * Create a Projection on top of TS that contains all columns from TS.
-   * Let rowIsDeleted the last column in the new Project.
-   * Enable fetching Deleted rows in TS.
-   * @param scan - TS to transform
-   * @return - new TS and a optionally a Project on top of it.
-   */
-  @Override
-  public RelNode visit(HiveTableScan scan) {
-    RelDataType tableRowType = scan.getTable().getRowType();
-    RelDataTypeField column = tableRowType.getField(
-        VirtualColumn.ROWISDELETED.getName(), false, false);
-    if (column == null) {
-      // This should not happen since Virtual columns are propagated for all native table scans in
-      // CalcitePlanner.genTableLogicalPlan()
-      throw new ColumnPropagationException("TableScan " + scan + " row schema does not contain " +
-          VirtualColumn.ROWISDELETED.getName() + " virtual column");
+    return dispatcher.invoke(relNode, new Context());
+  }
+
+  protected RelNode visitChild(RelNode parent, int i, RelNode child, Context context) {
+    stack.push(parent);
+    try {
+      RelNode newRel = dispatcher.invoke(child, context);
+      if (newRel != child) {
+        final List<RelNode> newInputs = new ArrayList<>(parent.getInputs());
+        newInputs.set(i, newRel);
+        return parent.copy(parent.getTraitSet(), newInputs);
+      }
+      return parent;
+    } finally {
+      stack.pop();
     }
+  }
+
+  protected RelNode visitChildren(RelNode rel, Context context) {
+    for (Ord<RelNode> input : Ord.zip(rel.getInputs())) {
+      rel = visitChild(rel, input.i, input.e, context);
+    }
+    return rel;
+  }
+
+  public static class Context {
```

Comment:
should it be public?

Reply:
It is public because it is exposed by the `visit` methods.